streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -509,38 +519,60 @@ void handleRevocation(final Collection<TopicPartition> revokedPartitions) {
             prepareCommitAndAddOffsetsToMap(commitNeededActiveTasks, consumedOffsetsPerTask);
         }
 
-        // even if commit failed, we should still continue and complete suspending those tasks,
-        // so we would capture any exception and throw
+        // even if commit failed, we should still continue and complete suspending those tasks, so we would capture
+        // any exception and rethrow it at the end
+        final Set<TaskId> corruptedTasks = new HashSet<>();
         try {
             commitOffsetsOrTransaction(consumedOffsetsPerTask);
+        } catch (final TaskCorruptedException e) {
+            log.warn("Some tasks were corrupted when trying to commit offsets, these will be cleaned and revived: {}",
+                     e.corruptedTasks());
+
+            // If we hit a TaskCorruptedException, we should just handle the cleanup for those corrupted tasks right here
+            corruptedTasks.addAll(e.corruptedTasks());
+            final Map<Task, Collection<TopicPartition>> corruptedTasksWithChangelogs = new HashMap<>();
+            for (final TaskId taskId : corruptedTasks) {
+                final Task task = tasks.task(taskId);
+                task.markChangelogAsCorrupted(task.changelogPartitions());
+                corruptedTasksWithChangelogs.put(task, task.changelogPartitions());
+            }
+            closeAndRevive(corruptedTasksWithChangelogs);
         } catch (final RuntimeException e) {
             log.error("Exception caught while committing those revoked tasks " + revokedActiveTasks, e);
+
+            // TODO: KIP-572 need to handle TimeoutException, may be rethrown from committing offsets under ALOS
```

Comment:
I don't think we need to handle `TimeoutException` here? If a client throws a `Timeout` it should be handled within `commitOffsetsOrTransaction` -- and `commitOffsetsOrTransaction` should only re-throw a timeout after `task.timeout.ms` expired and for this case we should indeed let the thread die.

Or actually, it seems we don't want `commitOffsetsOrTransaction` to catch client timeouts when we call it from here (we only want it to catch them during regular processing). Thus, we might need to pass in a boolean flag... And for this call, any client `TimeoutException` should be rethrown as `TaskCorruptedException` ?

Reply:
> commitOffsetsOrTransaction should only re-throw a timeout after task.timeout.ms expired

I don't think that's true -- for ALOS it just logs an error and then immediately rethrows

> any client TimeoutException should be rethrown as TaskCorruptedException

That's what I was thinking originally, but I wanted to keep this PR simple and not mess with any of the existing logic. For now we'll just let the thread die if we hit a TimeoutException when committing with ALOS from `handleRevocation` or `handleCorruption`. That seems good enough for 2.8, and we can follow up on the TimeoutException handling after this PR -- hence the TODO  🙂

Reply:
> I don't think that's true -- for ALOS it just logs an error and then immediately rethrows

Oh. We catch this one in `TaskManager#commit`...

> hence the TODO

Well, I was hoping to complete KIP-572 in 2.8...  So maybe we should just catch the `TimeoutException` here? Also, the change to pass in a boolean flag for EOS to rethrow `Timeout` directly seems to be rather simple?

Reply:
Ok I think we worked out a solution that addresses all the concerns and will have KIP-572 completed in 2.8: we always throw TimeoutException in ALOS, TaskCorruptedException in EOS. In both `handleCorrupted` and `handleRevocation` we catch these and then call `closeAndRevive` with a flag to indicate whether it should mark the changelogs as corrupted or not -- for ALOS tasks, this means the checkpoint will be maintained and the state does not have to be wiped out

Reply:
> Oh. We catch this one in TaskManager#commit...

Ah, sorry, I missed that. Yeah I guess that's true then, although we only go through TaskManager#commit in the `handleCorrupted` case. For `handleRevocation` we would just throw it directly -- and we can't go through TaskManager#commit without some further refactoring, for one thing because it just returns immediately without committing if we're in the middle of a rebalance.

In that case, we actually *should* let the TimeoutException kill the thread when it's thrown and caught during `handleCorruption`..? But for TimeoutException in `handleRevocation` we should actually invoke `maybeInitTaskTimeoutOrThrow` on all tasks, in addition to the other cleanup. 

Man I've been staring at this stuff too long 😖  Just lmk what you think of the latest

Reply:
> In that case, we actually should let the TimeoutException kill the thread when it's thrown and caught during handleCorruption..?

I guess we could do that; on the other hand, the task is corrupted already, and thus, it might also be ok to ignore `task.timeout.ms` and effectively reset the timeout? Or should we only reset the timeout _if_ the cleanup is successful? -- Don't have a strong opinion. Both solutions sound good to me.

> But for TimeoutException in handleRevocation we should actually invoke maybeInitTaskTimeoutOrThrow on all tasks, in addition to the other cleanup.

Similar to above: the task is getting revoked anyway and thus we could just ignore `task.timeout.ms` config for this case and just do the cleanup?

To me, `task.timeout.ms` applies mainly to the regular processing case. For task revocation/corruption it seem ok to ignore the timeout and just reset it.

Reply:
> To me, task.timeout.ms applies mainly to the regular processing case. For task revocation/corruption it seem ok to ignore the timeout and just reset it

Ok...I honestly didn't follow KIP-572 too closely but that's what I was thinking originally and then I got paranoid. But I managed the changes so I could just revert the last commit if that's how you responded. Sounds like that's the plan -- please give it one hopefully final pass once I do and we can finally wash our hands of this